src/backend/distributed/operations/shard_split.c

```diff
@@ -0,0 +1,1029 @@
+ * 'shardSplitPointsList'		: Split Points list for the source 'shardInterval'.
+ * 'nodeIdsForPlacementList'	: Placement list corresponding to split children.
+ */
+void
+SplitShard(SplitMode splitMode,
+		   SplitOperation splitOperation,
+		   uint64 shardIdToSplit,
+		   List *shardSplitPointsList,
+		   List *nodeIdsForPlacementList)
+{
+	if (XactModificationLevel > XACT_MODIFICATION_NONE)
+	{
+		ereport(ERROR, (errcode(ERRCODE_ACTIVE_SQL_TRANSACTION),
+						errmsg("cannot %s %s after other modifications "
+							   "in the same transaction.",
+							   SplitOperationName[splitOperation],
+							   SplitTargetName[splitOperation])));
+	}
+
+	ShardInterval *shardIntervalToSplit = LoadShardInterval(shardIdToSplit);
+	List *colocatedTableList = ColocatedTableList(shardIntervalToSplit->relationId);
+
+	/* sort the tables to avoid deadlocks */
+	colocatedTableList = SortList(colocatedTableList, CompareOids);
+	Oid colocatedTableId = InvalidOid;
+	foreach_oid(colocatedTableId, colocatedTableList)
+	{
+		/*
+		 * Block concurrent DDL / TRUNCATE commands on the relation. Similarly,
+		 * block concurrent citus_move_shard_placement() / isolate_tenant_to_new_shard()
+		 * on any shard of the same relation.
+		 */
+		LockRelationOid(colocatedTableId, ShareUpdateExclusiveLock);
+	}
+
+	ErrorIfCannotSplitShard(SHARD_SPLIT_API, shardIntervalToSplit);
+	ErrorIfCannotSplitShardExtended(
+		SHARD_SPLIT_API,
+		shardIntervalToSplit,
+		shardSplitPointsList,
+		nodeIdsForPlacementList);
+
+	List *workersForPlacementList = NIL;
+	Datum nodeId;
+	foreach_int(nodeId, nodeIdsForPlacementList)
+	{
+		uint32 nodeIdValue = DatumGetUInt32(nodeId);
+		WorkerNode *workerNode = LookupNodeByNodeId(nodeIdValue);
+
+		/* NodeId in Citus are unsigned and range from [1, 4294967296]. */
+		if (nodeIdValue < 1 || workerNode == NULL)
+		{
+			ereport(ERROR, (errcode(ERRCODE_ACTIVE_SQL_TRANSACTION),
+							errmsg("Invalid Node Id '%u'.", nodeIdValue)));
+		}
+
+		workersForPlacementList =
+			lappend(workersForPlacementList, (void *) workerNode);
+	}
+
+	if (splitMode == BLOCKING_SPLIT)
+	{
+		EnsureReferenceTablesExistOnAllNodesExtended(TRANSFER_MODE_BLOCK_WRITES);
+		BlockingShardSplit(
+			splitOperation,
+			shardIntervalToSplit,
+			shardSplitPointsList,
+			workersForPlacementList);
+	}
+	else
+	{
+		/* we only support blocking shard split in this code path for now. */
+		ereport(ERROR, (errmsg("Invalid split mode value %d.", splitMode)));
+	}
+}
+
+
+/*
+ * SplitShard API to split a given shard (or shard group) in blocking fashion
+ * based on specified split points to a set of destination nodes.
+ * 'splitOperation'             : Customer operation that triggered split.
+ * 'shardIntervalToSplit'       : Source shard interval to be split.
+ * 'shardSplitPointsList'		: Split Points list for the source 'shardInterval'.
+ * 'workersForPlacementList'	: Placement list corresponding to split children.
+ */
+static void
+BlockingShardSplit(SplitOperation splitOperation,
```

Comment:
It seems like we even block `SELECT`s at some point -- probably end of the shard split. Is that expected that SELECTs are also blocked? Could be because of some code that this PR relies on, but probably needs some further look.

```SQL
select * from citus_lock_waits;
┌─[ RECORD 1 ]──────────────────────────┬──────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ waiting_gpid                          │ 10000056139                                                                                          │
│ blocking_gpid                         │ 10000055984                                                                                          │
│ blocked_statement                     │ select count(*) from test ;                                                                          │
│ current_statement_in_blocking_process │ WITH shard_ranges AS (SELECT ((1879048192 + i))::text as r, 7 as n FROM generate_series(0,13421) i) ↵│
│                                       │ SELECT citus_split_shard_by_split_points(102038, array_agg(r), array_agg(n) || 7) FROM shard_ranges; │
│ waiting_nodeid                        │ 1                                                                                                    │
│ blocking_nodeid                       │ 1                                                                                                    │
└───────────────────────────────────────┴──────────────────────────────────────────────────────────────────────────────────────────────────────┘

Time: 371.462 ms

```

Looking closely, the SELECT is blocked on accesssharelock on the worker:
```SQL
 select relation::regclass, mode from pg_locks where pid = 71890 and not granted;
┌─[ RECORD 1 ]───────────────┐
│ relation │ test_102038     │
│ mode     │ AccessShareLock │
└──────────┴─────────────────┘

Time: 2.255 ms
```

Because at some point we acquire `AccessExclusiveLock` on the shard we are splitting:
```SQL
select relation::regclass, mode from pg_locks where pid = 56076 and granted and relation is not null;
┌─[ RECORD 1 ]────────────────────────┐
│ relation │ pg_dist_node             │
│ mode     │ AccessShareLock          │
├─[ RECORD 2 ]────────────────────────┤
│ relation │ pg_dist_placement        │
│ mode     │ AccessShareLock          │
├─[ RECORD 3 ]────────────────────────┤
│ relation │ pg_dist_placement        │
│ mode     │ RowExclusiveLock         │
├─[ RECORD 4 ]────────────────────────┤
│ relation │ pg_dist_partition        │
│ mode     │ AccessShareLock          │
├─[ RECORD 5 ]────────────────────────┤
│ relation │ pg_dist_shard            │
│ mode     │ AccessShareLock          │
├─[ RECORD 6 ]────────────────────────┤
│ relation │ pg_dist_shard            │
│ mode     │ RowExclusiveLock         │
├─[ RECORD 7 ]────────────────────────┤
│ relation │ test_102038              │
│ mode     │ AccessShareLock          │
├─[ RECORD 8 ]────────────────────────┤
│ relation │ test_102038              │
**│ mode     │ AccessExclusiveLock      │**
├─[ RECORD 9 ]────────────────────────┤
│ relation │ test                     │
│ mode     │ ShareUpdateExclusiveLock │
└──────────┴──────────────────────────┘

Time: 1.866 ms
```

Also, the blocked `SELECT` fails at the end of the tx;
```SQL
select count(*) from test ;
ERROR:  relation "public.test_102038" does not exist
CONTEXT:  while executing command on localhost:9702
```

Reply:
that's because we drop the shards immediately, we should implement deferred drop

Reply:
Thanks for the reply Marco. @onderkalaci, This is the first MVP PR, deferred drop will be implemented in a coming PR.

Reply:
Marking this as resolved. @onderkalaci , please feel free to reopen if you believe this is not closed issue.